pkg/util/reflector/prometheus/prometheus.go

```diff
@@ -0,0 +1,116 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package prometheus sets the cache DefaultMetricsFactory to produce
+// prometheus metrics. To use this package, you just have to import it.
+
+package prometheus
+
+import (
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+func init() {
+	cache.SetReflectorMetricsProvider(prometheusMetricsProvider{})
+}
+
+type prometheusMetricsProvider struct{}
+
+func (_ prometheusMetricsProvider) NewListsMetric(name string) cache.CounterMetric {
+	lists := prometheus.NewCounter(prometheus.CounterOpts{
+		Subsystem: name,
+		Name:      "lists_total",
+		Help:      "Total number of API lists done by the reflector: " + name,
+	})
+	prometheus.MustRegister(lists)
+	return lists
+}
+
+// use summary to get averages and percentiles
+func (_ prometheusMetricsProvider) NewListDurationMetric(name string) cache.SummaryMetric {
+	listDuration := prometheus.NewSummary(prometheus.SummaryOpts{
+		Subsystem: name,
+		Name:      "list_duration",
+		Help:      "How long an API list takes to return and decode for reflector: " + name,
+	})
+	prometheus.MustRegister(listDuration)
+	return listDuration
+}
+
+// use summary to get averages and percentiles
+func (_ prometheusMetricsProvider) NewItemsInListMetric(name string) cache.SummaryMetric {
+	itemsPerList := prometheus.NewSummary(prometheus.SummaryOpts{
+		Subsystem: name,
+		Name:      "items_per_list",
+		Help:      "How many items an API list returns to reflector: " + name,
+	})
+	prometheus.MustRegister(itemsPerList)
+	return itemsPerList
+}
+
+func (_ prometheusMetricsProvider) NewWatchesMetric(name string) cache.CounterMetric {
+	watches := prometheus.NewCounter(prometheus.CounterOpts{
+		Subsystem: name,
+		Name:      "watches_total",
+		Help:      "Total number of API watches done by the reflector: " + name,
+	})
+	prometheus.MustRegister(watches)
+	return watches
+}
+
+func (_ prometheusMetricsProvider) NewShortWatchesMetric(name string) cache.CounterMetric {
+	watches := prometheus.NewCounter(prometheus.CounterOpts{
+		Subsystem: name,
+		Name:      "short_watches_total",
+		Help:      "Total number of short API watches done by the reflector: " + name,
+	})
+	prometheus.MustRegister(watches)
+	return watches
+}
+
+// use summary to get averages and percentiles
+func (_ prometheusMetricsProvider) NewWatchDurationMetric(name string) cache.SummaryMetric {
+	watchDuration := prometheus.NewSummary(prometheus.SummaryOpts{
+		Subsystem: name,
+		Name:      "watch_duration",
```

Comment:
Prometheus convention is to do seconds, and to include that in the name i.e. `watch_duration_seconds`.  Same for list duration.